### PR TITLE
Search bar for mapping missions

### PR DESCRIPTION
**Pull request by Jorque**
Adds a search bar to the mapping missions selection activity that allows user to search for specific mapping missions and display them. At the moment, the search is case sensitive and the search will output all missions whose name starts with the input.

### CODE REVIEW

File: app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt [hunk 1]
```diff
@@ -128,10 +140,76 @@ class FirebaseMappingMissionDao @Inject constructor(
...
+        val query =
+            if (ownerUid != null) privateFilteredMappingMissionsQuery else sharedFilteredMappingMissionsQuery
+        val listener =
+            if (ownerUid != null) privateFilteredMappingMissionsListener else sharedFilteredMappingMissionsListener
+
+        if (query != null && listener != null) {
+            query.removeEventListener(listener)
+        }
+
+        val missionsQuery = rootRef.orderByChild("nameUpperCase").startAt(filter).endAt(filter + "\uf8ff")
```
Comment (GabrielFleischer) [inline app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt:170]: What is "\uf8ff" ?

File: app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt [hunk 2]
```diff
@@ -128,10 +140,76 @@ class FirebaseMappingMissionDao @Inject constructor(
...
+            privateFilteredMappingMissionsListener = missionsListener
+        } else {
+            sharedFilteredMappingMissionsFilter = filter
+            sharedFilteredMappingMissionsQuery = missionsQuery
+            sharedFilteredMappingMissionsListener = missionsListener
+        }
+    }
+
+    override fun updatePrivateFilteredMappingMissions(ownerUid: String, filter: String?) {
+        if (filter != null && privateFilteredMappingMissionsFilter != filter.toUpperCase()) {
```
Comment (GabrielFleischer) [inline app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt:197]: Cirrus raised a warning here with toUpperCase().
This hould maybe be
```suggestion
        if (filter != null && privateFilteredMappingMissionsFilter != filter.toUpperCase(Locale.ROOT)) {
```

File: app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt [hunk 3]
```diff
@@ -128,10 +140,76 @@ class FirebaseMappingMissionDao @Inject constructor(
...
+            updateFilteredMappingMissions(ownerUid, filter.toUpperCase())
```
Comment (GabrielFleischer) [inline app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt:198]: Same here

File: app/src/main/java/ch/epfl/sdp/drone3d/ui/mission/MappingMissionSelectionActivity.kt [hunk 4]
```diff
@@ -36,55 +38,131 @@ class MappingMissionSelectionActivity : AppCompatActivity() {
...
+        val checked: Boolean,
+        val sharedVisible: Boolean,
+        val privateVisible: Boolean
+    ) {
         PRIVATE(true, false, true),
         SHARED(false, true, false)
     }
 
-    private val currentType = MutableLiveData(StorageType.PRIVATE)
+    private val currentType = MutableLiveData(Pair<StorageType, String?>(StorageType.PRIVATE, null))
```
Comment (GabrielFleischer) [inline app/src/main/java/ch/epfl/sdp/drone3d/ui/mission/MappingMissionSelectionActivity.kt:50]: As this is more than just the type, maybe rename the value to something more meaningful

**Review by GabrielFleischer — APPROVED** [review]

File: app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt (hunk 1 above)
Comment (Jorque) [reply app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt:170]: \uf8ff is the last unicode character. I use it so the query will get all missions that start with the substring filter

File: app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt (hunk 2 above)
Comment (Jorque) [reply app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt]: Done

File: app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt (hunk 3 above)
Comment (Jorque) [reply app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/FirebaseMappingMissionDao.kt]: Done

File: app/src/main/java/ch/epfl/sdp/drone3d/ui/mission/MappingMissionSelectionActivity.kt (hunk 4 above)
Comment (Jorque) [reply app/src/main/java/ch/epfl/sdp/drone3d/ui/mission/MappingMissionSelectionActivity.kt]: Done

**Review by jonasblanc — APPROVED** [review]
After testing I realize that for exemple "mission 5" comes after "mission 10" when you search "mission". If you can think of an easy way to fix that, it would be nice, otherwise we can keep that fix for later.

File: app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/MappingMissionDao.kt [hunk 5]
```diff
@@ -34,6 +34,32 @@ interface MappingMissionDao {
      */
     fun getSharedMappingMissions(): LiveData<List<MappingMission>>
 
+    /**
+     * Returns the list of all private mapping missions filtered by the current filter,
+     * this live data is updated in continuous time if the private missions are modified
+     * and also when updatePrivateFilteredMappingMissions is called with another filter
+     */
+    fun getPrivateFilteredMappingMissions(): LiveData<List<MappingMission>>
```
[the remark this answers is not in the document]
Comment (Jorque) [reply app/src/main/java/ch/epfl/sdp/drone3d/service/storage/dao/MappingMissionDao.kt:43]: Done

**Pull request merged**